File: CppFd/RemoveNthNodeFromEndOfList.cpp
#include <iostream>
#include <vector>
#include "Solution.cpp"
#include "ListNode.h"

using namespace std;

class RemoveNthNodeFromEndOfList {
public:
    struct TestCase {
        ListNode* head;
        int n;
        ListNode* expected;
        TestCase(ListNode* h, int num, ListNode* exp) : head(h), n(num), expected(exp) {}
    };

    static ListNode* arrayToList(const vector<int>& arr) {
        if (arr.empty()) return nullptr;
        ListNode* head = new ListNode(arr[0]);
        ListNode* current = head;
        for (size_t i = 1; i < arr.size(); i++) {
            current->next = new ListNode(arr[i]);
            current = current->next;
        }
        return head;
    }

    static vector<int> listToArray(ListNode* head) {
        vector<int> arr;
        while (head) {
            arr.push_back(head->val);
            head = head->next;
        }
        return arr;
    }

    vector<TestCase> getTestCases() {
        return {
            {arrayToList({1,2,3,4,5}), 2, arrayToList({1,2,3,5})},
            {arrayToList({1}), 1, arrayToList({})},
            {arrayToList({1,2}), 1, arrayToList({1})},
            {arrayToList({1,2,3,4,5,6}), 3, arrayToList({1,2,3,5,6})},
            {arrayToList({10,20,30,40,50}), 5, arrayToList({20,30,40,50})},
            {arrayToList({7,8,9,10,11,12,13}), 1, arrayToList({7,8,9,10,11,12})},
            {arrayToList({5,10,15,20}), 2, arrayToList({5,10,20})},
            {arrayToList({100,200,300}), 3, arrayToList({200,300})},
            {arrayToList({9,8,7,6,5,4,3,2,1}), 4, arrayToList({9,8,7,6,5,3,2,1})},
            {arrayToList({1,2,3,4,5}), 5, arrayToList({2,3,4,5})}
        };
    }

    void runTests() {
        vector<TestCase> testCases = getTestCases();
        for (size_t i = 0; i < testCases.size(); i++) {
            Solution solution;
            vector<int> actual = listToArray(solution.removeNthFromEnd(testCases[i].head, testCases[i].n));
            vector<int> expected = listToArray(testCases[i].expected);

            if (actual != expected) {
                cout << "Wrong Answer : " << i << "/" << testCases.size() << endl;
                return;
            }
        }

        cout << "Accepted" << endl;
    }
};

int main() {
    RemoveNthNodeFromEndOfList tester;
    tester.runTests();
    return 0;
}

File: JavaFd/RemoveNthNodeFromEndOfList.java
import java.util.*;

public class RemoveNthNodeFromEndOfList {
    public static class TestCase {
        ListNode head;
        int n;
        ListNode expected;

        public TestCase(ListNode head, int n, ListNode expected) {
            this.head = head;
            this.n = n;
            this.expected = expected;
        }
    }

    private static ListNode arrayToList(int[] arr) {
        if (arr.length == 0) return null;
        ListNode head = new ListNode(arr[0]);
        ListNode current = head;
        for (int i = 1; i < arr.length; i++) {
            current.next = new ListNode(arr[i]);
            current = current.next;
        }
        return head;
    }

    private static int[] listToArray(ListNode head) {
        List<Integer> list = new ArrayList<>();
        while (head != null) {
            list.add(head.val);
            head = head.next;
        }
        return list.stream().mapToInt(i -> i).toArray();
    }

    public static TestCase[] getTestCases() {
        return new TestCase[]{
            new TestCase(arrayToList(new int[]{1,2,3,4,5}), 2, arrayToList(new int[]{1,2,3,5})),
            new TestCase(arrayToList(new int[]{1}), 1, arrayToList(new int[]{})),
            new TestCase(arrayToList(new int[]{1,2}), 1, arrayToList(new int[]{1})),
            new TestCase(arrayToList(new int[]{1,2,3,4,5,6}), 3, arrayToList(new int[]{1,2,3,5,6})),
            new TestCase(arrayToList(new int[]{10,20,30,40,50}), 5, arrayToList(new int[]{20,30,40,50})),
            new TestCase(arrayToList(new int[]{7,8,9,10,11,12,13}), 1, arrayToList(new int[]{7,8,9,10,11,12})),
            new TestCase(arrayToList(new int[]{5,10,15,20}), 2, arrayToList(new int[]{5,10,20})),
            new TestCase(arrayToList(new int[]{100,200,300}), 3, arrayToList(new int[]{200,300})),
            new TestCase(arrayToList(new int[]{9,8,7,6,5,4,3,2,1}), 4, arrayToList(new int[]{9,8,7,6,5,3,2,1})),
            new TestCase(arrayToList(new int[]{1,2,3,4,5}), 5, arrayToList(new int[]{2,3,4,5}))
        };
    }

    public static void main(String[] args) {
        TestCase[] testCases = getTestCases();
        boolean success = true;
        int idx = -1;
        for (int i = 0; i < testCases.length; i++) {
            Solution obj = new Solution();
            int[] actual = listToArray(obj.removeNthFromEnd(testCases[i].head, testCases[i].n));
            int[] expected = listToArray(testCases[i].expected);
            if (!Arrays.equals(actual, expected)) {
                success = false;
                idx = i;
                break;
            }
        }
        if (success) {
            System.out.println("Accepted");
        } else {
            System.out.println("Wrong Answer : " + idx + "/" + testCases.length);
        }
    }
}